### ISSUE TITLE
Invalid MP-JWT TCK test cases

### ISSUE THREAD

**Issue by chunlongliang-ibm** [eclipse/microprofile-jwt-auth#104]
1. The test org.eclipse.microprofile.jwt.tck.config.IssValidationFailTest checks for a property that is no longer part of the spec  mp.jwt.verify.requireiss=false.  

2. The test org.eclipse.microprofile.jwt.tck.config.PublicKeyAsPEMTest  uses a Pem key without the required ---BEGIN--- , which is not in PKCS#8 PEM key format

3. The tests PublicKeyAsJWKLocationURLTest and PublicKeyAsPEMLocationURLTest contain a hardcoded URL for the app  that is incompatible with other container defaults.  We have to change the test URL with system property or environment variable.

4. The tests PublicKeyASJWKLocationTest and PublicKeyAsJWKSLocationTest use a public key location property with a value of "classpath:/signer-key4k.jwk"  or something similar.  The "classpath" prefix is not part of the spec, and should be considered as custom URL protocol.  

5. There are quite a few older tests that contains web-inf/classes/publicKey.pem without  mp.jwt.verify.publickey.location defined. Note: We have to define mp.jwt.verify.publickey.location for the test.

6. Some tests contain  pkcs#1 format keys. Note, we change our code to tolerate pkcs#1 PEM key

**Comment by arjantijms** [comment]
See also this PR https://github.com/eclipse/microprofile-jwt-auth/pull/102

**Comment by starksm64** [comment]
What I had to do was to replace the CDI java.security.Principal bean implementation for this to work with the Weld/WildflySwarm(now Thorntail) implementation. If this is not a generally applicable approach, we can just have the test validate that the Principal#getName() value maps to what is expected based on the underlying JWT. I see in the PR that the injection is being changed to a SecurityContext. The point of the test was to ensure that supporting injection of JsonWebToken did not break injection of the Principal type.

There are a few other challenges that have been raised, so it looks like we'll have to create a TCK patch release to incorporate changes to improve the TCK to run across multiple implementations.

**Comment by arjantijms** [comment]
@starksm64 I agree that for this specific test it's indeed not the best approach to inject a security context and test that. The test should indeed test for the injection of Principal directly.

I think it does reveal the need for two tests:

1. Testing that injection of Principal works, and that it has the correct name.
2. Testing that the Principal obtained from SecurityContext can be downcasted correctly

>What I had to do was to replace the CDI java.security.Principal bean implementation for this to work with the Weld/WildflySwarm(now Thorntail) implementation

By doing so, did you still make it a normal scoped proxy then? Otherwise if people inject it into say an application scoped bean different users will see each other's principal (depending on which one would be injected first)

**Comment by starksm64** [comment]
@arjantijms, sorry I meant for that comment to be on your #103 issue, but yes, the principal replacement should be local to the war, but I did not test for side effects. What your proposing for the split of the test into two is what I'll do when I merge the #102 PR.

**Comment by starksm64** [comment]
@chunlongliang-ibm, on your 5) issue, those are the 1.0 tests that were not updated to include mp config properties as a test of allowing vendor specific settings to work as they did in 1.0. The expectation was that if a test war did not contain a META-INF/microprofile-config.properties resource, the vendor test harness adaptor would setup the application deployment as it did for the 1.0 TCK. This way, legacy 1.0 MP-JWT deployments should work with a 1.1 MP-JWT runtime.

**Comment by starksm64** [comment]
@chunlongliang-ibm, on your 4) issue, that was an artifact from when we were discussing that as a supported URL scheme to support, but this does need to be reverted.

**Comment by arjantijms** [comment]
@starksm64 
>sorry I meant for that comment to be on your #103 issue, 

No problems, I understood that ;)

>but yes, the principal replacement should be local to the war, but I did not test for side effects. 

I took a look at this, but it looks troublesome:

https://github.com/thorntail/thorntail/blob/master/fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/deployment/auth/cdi/PrincipalProducer.java

```java
@Priority(1)
@Alternative
public class PrincipalProducer {

    /**
     * The producer method for the current JsonWebToken
     *
     * @return
     */
    @Produces
    @Dependent
    JsonWebToken currentJWTPrincipalOrNull() {
        return MPJWTProducer.getJWTPrincpal();
    }
}
```

What you're doing here is returning a dependent scoped instance, i.e. an instance without a proxy. 

This instance will not reflect the current state of authentication once injected. For instance, if you inject this into a session scoped bean, and the caller authenticates as a different caller in the next request, the session scoped bean will still have the previous instance.

Even worse is when you inject this into an application scoped bean. This will then forever see the first principal injected. It also won't work with @RunAs, and mid request logout (calling `HttpServletRequest.logout()`). This is a very big security risk.

I noticed the implementation code in Thorntail and SmallRye uses a lot of custom code and overrides for everything (own JAX-RS SecurityContext wrapper, this Bean alternative, etc). In general IMHO this should be avoided, and it's best to make use of the existing security primitives. Using those, you simply provide the correct principal once to the server, and then the server makes that principal available everywhere, and takes care of the scoping (makes sure it corresponds to the internal security context of the server).

I did this myself as follows here:

https://github.com/payara/Payara/blob/master/appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/eesecurity/JWTAuthenticationMechanism.java#L80

The principal is handed over to the container and that's it. It makes it available everywhere where its needed, including propagation to EJB and what have you, without needing any additional custom wrappers or alternatives for that. You also don't need any leak detection or cleanup since that's what the server does already for its 'normal' security.

I used the Java EE security APIs above (as an implementation detail), but for EE 7 compatibility you can use JASPIC just as well (again, as implementation detail), or for a JBoss native version even the Undertow APIs for this directly (which is what the thin JASPIC layer in Undertow ultimately calls anyway).

Just my 2 cents ;)

**Comment by arjantijms** [comment]
@starksm64 

Looking a bit more at the repo, I see there's actually an (Undertow) authentication mechanism being used that "officially" installs the principal:

https://github.com/thorntail/thorntail/blob/master/fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/deployment/auth/JWTAuthMechanism.java#L84

I must have been confused with SmallRye then, where things such as this are used: https://github.com/smallrye/smallrye-jwt/blob/master/implementation/src/main/java/io/smallrye/jwt/auth/jaxrs/JWTSecurityContext.java

**Comment by brutif** [comment]
@starksm64 
regarding item 3, the tests need something like this: 
```
        String portProp = System.getenv("testport");
        String ctxRootProp = System.getenv("testContextRoot");
        String port = portProp != null ? portProp : "8080";
        String ctxRoot = ctxRootProp != null ? ctxRootProp + "/" : "";
        String keyLoc = "http://localhost:" + port + "/" + ctxRoot + "jwks/endp/publicKey4kAsJWKS?kid=publicKey4k";
configProps.setProperty(Names.VERIFIER_PUBLIC_KEY_LOCATION, keyLoc);
```

**Comment by brutif** [comment]
@starksm64 
update on above, - Now that I think about it, System.getProperty is probably better, as java can't do setenv on the fly.

**Comment by starksm64** [comment]
I have merged fixes/updates for all of these issues. I'll look to do a 1.1.1-RC1 release early this week so these can be tested.

**Comment by brutif** [comment]
@starksm64 Thanks for the quick updates!    I tried the new updates today and saw just a couple of minor things 
- the test jar isn't picking up the new util package
- the test jar says v1.2 
- publicKeyAsPEMLocationURLTest#114 still has hardcoded URL.
